src/qnngds/utilities.py

```diff
@@ -174,6 +177,191 @@ def offset(overlap_port):
     return DIE
 
 
+def calculate_available_space_for_dev(
+    die_size: Tuple[Union[int, float], Union[int, float]] = (dflt.die_w, dflt.die_w),
+    pad_size: Tuple[Union[int, float], Union[int, float]] = dflt.pad_size,
+    contact_l: Union[int, float] = dflt.ebeam_overlap,
+    isolation: Union[int, float] = dflt.die_outline,
+) -> Tuple[float, float]:
+    """Calculates the maximum space available for a device in a die_cell.
+
+    Note that the device should be even smaller than this function's output, as
+    it does not account for routing from the device to the die's ports.
+
+    Parameters:
+        die_size (tuple of int or float): Overall size of the cell (width, height).
+        pad_size (tuple of int or float): Dimensions of the cell's pads (width, height).
+        contact_l (int or float): Extra length of the routes above the ports to
+            assure alignment with the device (useful for ebeam lithography).
+        isolation (int or float): The width of the pads outline.
+
+    Returns:
+        Tuple[float, float]: A tuple containing the width and height of space
+        available for a device in a die_cell
+    """
+
+    dev_max_size = [
+        2 * (die_w / 2 - 3 * isolation - pad_size[1] - 2 * contact_l)
+        for die_w in die_size
+    ]
+    return dev_max_size[0], dev_max_size[1]
+
+
+def calculate_contact_w(
+    circuit_ports: List[Port] = [Port(width=50)],
+    overlap_w: Union[int, float] = dflt.ebeam_overlap,
+) -> Union[int, float]:
+    """Calculate the minimal width acceptable for the contact from the die_cell
+    to the circuit/experiment.
+
+    The conditions are that the contact width should be bigger than (1) the
+    biggest port of the circuit, (2) the overlap_w (to avoid a short).
+
+    Parameters:
+        circuit_ports (list of Port): The ports of the circuit to be placed in the
+            cell (use .get_ports()).
+        overlap_w (int or float): The overlap width in µm (accounts for
+            misalignment between 1st and 2nd ebeam exposures).
+
+    Returns:
+        (int or float): the suggested contact_w to input in the die_cell of the
+        given circuit
+    """
+    max_circuit_port_width = max([port.width for port in circuit_ports])
+    return max(max_circuit_port_width, overlap_w)
+
+
+def add_optimalStep_to_dev(
```

Comment:
I think PEP recommends `snake_case` for all class methods, functions, and variables. Maybe `add_optimalstep_to_dev` or `add_optimal_step_to_dev`?
Also this function seems very nice